src/wasm/wasm-emscripten.cpp

```diff
@@ -308,9 +310,14 @@ void EmscriptenGlueGenerator::generateJSCallThunks(
 std::vector<Address> getSegmentOffsets(Module& wasm) {
   std::vector<Address> segmentOffsets;
   for (unsigned i = 0; i < wasm.memory.segments.size(); ++i) {
-    Const* addrConst = wasm.memory.segments[i].offset->cast<Const>();
-    auto address = addrConst->value.geti32();
-    segmentOffsets.push_back(address);
+    if (auto* addrConst = wasm.memory.segments[i].offset->dynCast<Const>()) {
+      auto address = addrConst->value.geti32();
+      segmentOffsets.push_back(address);
+    } else {
+      // TODO(sbc): Wasm shared libraries have data segments with non-const
```

Comment:
In theory this may not have to change, as everything will be indexed relative to the start of the segment, and so 0 is roughly exactly correct, with respect to how we actually use this, i.e. reading things out of the memory statically.
But we can yank this TODO once we're more sure that's the case (i.e. once we actually support EM_ASM in shared libraries), just thinking out loud here.